clouddriver-docker/src/test/groovy/com/netflix/spinnaker/clouddriver/docker/registry/provider/agent/DockerRegistryImageCachingAgentTest.groovy

```diff
@@ -0,0 +1,113 @@
+import com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryTags
+import com.netflix.spinnaker.clouddriver.docker.registry.security.DockerRegistryCredentials
+import spock.lang.Specification
+
+import java.time.Instant
+
+class DockerRegistryImageCachingAgentTest extends Specification {
+
+  DockerRegistryImageCachingAgent agent
+  def credentials = Mock(DockerRegistryCredentials)
+  def provider = Mock(DockerRegistryCloudProvider)
+  def client = Mock(DockerRegistryClient)
+
+  def setup() {
+    credentials.client >> client
+    agent = new DockerRegistryImageCachingAgent(provider, "test-docker", credentials, 0, 1, 1, "test-registry")
+  }
+
+  def "tags loaded from docker registry should be cached"() {
+    given:
+    credentials.repositories >> repositories
+    def total = 0
+    for (def i = 0; i < repositories.size(); i++) {
+      client.getTags(repositories[i]) >> new DockerRegistryTags().tap {
+        name=repositories[i]
+        tags=repoTags[i]
+      }
+      total += repoTags[i].size()
+    }
+
+    when:
+    def cacheResult = agent.loadData(null)
+
+    then:
+    cacheResult.cacheResults.get("taggedImage").size() == total
+    cacheResult.cacheResults.get("imageId").size() == total
+    for (def i = 0; i < total; i++) {
+      cacheResult.cacheResults.get("taggedImage")[i].getId() != ""
+      cacheResult.cacheResults.get("imageId")[i].getId() != ""
+      def repoAndTag = cacheResult.cacheResults.get("taggedImage")[i].attributes.get("name").split(":")
+      repoTags[repositories.indexOf(repoAndTag[0])].contains(repoAndTag[1])
+      cacheResult.cacheResults.get("taggedImage")[i].attributes.get("digest") == null
+      cacheResult.cacheResults.get("taggedImage")[i].attributes.get("date") == null
+    }
+
+    where:
+    repositories          | repoTags
+    ["repo-1"]            | [["tag-1", "tag-2"]]
+    ["repo-1", "repo-2" ] | [["tag-1-1"], ["tag-2-1", "tag-2-2"]]
+  }
+
+  def "cached tags should include creation date"() {
+    given:
+    credentials.sortTagsByDate >> true
+    credentials.repositories >> ["repo-1"]
+    client.getTags("repo-1") >> new DockerRegistryTags().tap {
+      name="repo-1"
+      tags=["tag-1", "tag-2"]
+    }
+    client.getCreationDate(*_) >> Instant.EPOCH
+
+    when:
+    def cacheResult = agent.loadData(null)
+
+    then:
+    cacheResult.cacheResults.get("taggedImage").size() == 2
+    cacheResult.cacheResults.get("imageId").size() == 2
+    cacheResult.cacheResults.get("taggedImage")[0].attributes.get("date") == Instant.EPOCH
+    cacheResult.cacheResults.get("taggedImage")[1].attributes.get("date") == Instant.EPOCH
+  }
+
+  def "cached tags should include digest"() {
+    given:
+    credentials.trackDigests >> true
+    credentials.repositories >> ["repo-1"]
+    client.getTags("repo-1") >> new DockerRegistryTags().tap {
+      name="repo-1"
+      tags=["tag-1", "tag-2"]
+    }
+    client.getDigest(*_) >> "123"
```

Comment:
It might be a bit more robust to return different digests for the different tags, and ensure the digests appear correctly in the output.